// redstone-smartweave/lib/cjs/plugins/Evolve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Evolve = void 0;
const _smartweave_1 = require("..");
/*
...I'm still not fully convinced to the whole "evolve" idea.

IMO It makes it very hard to audit what exactly the smart contract's code at given txId is doing (as it requires
to analyse its whole interactions history and verify if some of them do not modify original contract's source code).

IMO instead of using "evolve" feature - a new contract version should be deployed (with "output state"
from previous version set as "input state" for the new version).

Instead of using "evolve" feature - one could utilise the "contracts-registry" approach:
https://github.com/redstone-finance/redstone-smartweave-contracts/blob/main/src/contracts-registry/contracts-registry.contract.ts
https://viewblock.io/arweave/address/XQkGzXG6YknJyy-YbakEZvQKAWkW2_aPRhc3ShC8lyA?tab=state
- it keeps track of all the versions of the given contract and allows to retrieve the latest version by contract's "business" name -
without the need of hard-coding contract's txId in the client's source code.

This also makes it easier to audit given contract - as you keep all its versions in one place.
*/
function isEvolveCompatible(state) {
    if (!state) {
        return false;
    }
    const settings = evalSettings(state);
    return state.evolve !== undefined || settings.has('evolve');
}
class Evolve {
    constructor(definitionLoader, executorFactory) {
        this.definitionLoader = definitionLoader;
        this.executorFactory = executorFactory;
        this.logger = _smartweave_1.LoggerFactory.INST.create('Evolve');
        this.modify = this.modify.bind(this);
    }
    async modify(state, executionContext) {
        const contractTxId = executionContext.contractDefinition.txId;
        this.logger.debug(`trying to evolve for: ${contractTxId}`);
        if (!isEvolveCompatible(state)) {
            this.logger.debug('State is not evolve compatible');
            return executionContext;
        }
        const currentSrcTxId = executionContext.contractDefinition.srcTxId;
        const settings = evalSettings(state);
        // note: from my understanding - this variable holds the id of the transaction with updated source code.
        const evolve = state.evolve || settings.get('evolve');
        let canEvolve = state.canEvolve || settings.get('canEvolve');
        // By default, contracts can evolve if there's not an explicit `false`.
        if (canEvolve === undefined || canEvolve === null) {
            canEvolve = true;
        }
        if (evolve && /[a-z0-9_-]{43}/i.test(evolve) && canEvolve) {
            this.logger.debug('Checking evolve:', {
                current: currentSrcTxId,
                evolve
            });
            if (currentSrcTxId !== evolve) {
                try {
                    // note: that's really nasty IMO - loading original contract definition,
                    // but forcing different sourceTxId...
                    this.logger.info('Evolving to: ', evolve);
                    const newContractDefinition = await this.definitionLoader.load(contractTxId, evolve);
                    const newHandler = (await this.executorFactory.create(newContractDefinition));
                    //FIXME: side-effect...
                    executionContext.contractDefinition = newContractDefinition;
                    executionContext.handler = newHandler;
                    this.logger.debug('evolved to:', {
                        evolve: evolve,
                        newSrcTxId: executionContext.contractDefinition.srcTxId,
                        current: currentSrcTxId,
                        txId: executionContext.contractDefinition.txId
                    });
                    return executionContext;
                }
                catch (e) {
                    throw new _smartweave_1.SmartWeaveError(_smartweave_1.SmartWeaveErrorType.CONTRACT_NOT_FOUND, {
                        message: `Contract having txId: ${contractTxId} not found`,
                        requestedTxId: contractTxId
                    });
                }
            }
        }
        return executionContext;
    }
}
exports.Evolve = Evolve;
function evalSettings(state) {
    // default  - empty
    let settings = new Map();
    if (state.settings) {
        // for Iterable format
        if (isIterable(state.settings)) {
            settings = new Map(state.settings);
            // for Object format
        }
        else if (isObject(state.settings)) {
            settings = new Map(Object.entries(state.settings));
        }
    }
    return settings;
}
function isIterable(obj) {
    // checks for null and undefined
    if (obj == null) {
        return false;
    }
    return typeof obj[Symbol.iterator] === 'function';
}
function isObject(obj) {
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}
