// redstone-smartweave/lib/cjs/core/SmartWeaveBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartWeaveBuilder = void 0;
const _smartweave_1 = require("..");
class SmartWeaveBuilder {
    constructor(_arweave) {
        this._arweave = _arweave;
    }
    setDefinitionLoader(value) {
        this._definitionLoader = value;
        return this;
    }
    setInteractionsLoader(value) {
        this._interactionsLoader = value;
        return this;
    }
    setInteractionsSorter(value) {
        this._interactionsSorter = value;
        return this;
    }
    setExecutorFactory(value) {
        this._executorFactory = value;
        return this;
    }
    setStateEvaluator(value) {
        this._stateEvaluator = value;
        return this;
    }
    overwriteSource(sourceCode) {
        if (this._executorFactory == null) {
            throw new Error('Set base ExecutorFactory first');
        }
        this._executorFactory = new _smartweave_1.DebuggableExecutorFactory(this._executorFactory, sourceCode);
        return this.build();
    }
    build() {
        return new _smartweave_1.SmartWeave(this._arweave, this._definitionLoader, this._interactionsLoader, this._interactionsSorter, this._executorFactory, this._stateEvaluator);
    }
}
exports.SmartWeaveBuilder = SmartWeaveBuilder;
